### PR TITLE
Initial reconnection facilities for Tcp and Http

### PR DESCRIPTION
**Pull request by ktoso**
These impls are a bit tied to our use-cases, could be made cleaner later.

### CODE REVIEW

File: src/main/scala/akka/contrib/stream/pattern/reconnect/Reconnecter.scala [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+import akka.actor._
+import Reconnecter._
+import akka.stream.FlowMaterializer
+import akka.stream.stage.{ Context, Directive, PushStage, TerminationDirective }
+import akka.util.ByteString
+
+import scala.concurrent.duration.FiniteDuration
+
+object Reconnecter {
```
Comment (hseeberger) [inline src/main/scala/akka/contrib/stream/pattern/reconnect/Reconnecter.scala]: Maybe Reconnector?

File: src/main/scala/akka/contrib/stream/pattern/reconnect/Reconnecter.scala [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+  private[reconnect] case object Reconnect
+}
+
+/**
+ * Contains logic of issuing connect() calls within expected intervals if a connection terminates
+ */
+abstract class Reconnecter[T, D <: ReconnectionData[T]](initialData: D)
+    extends Actor with ActorLogging {
+
+  implicit val dispatcher = context.dispatcher
```
Comment (hseeberger) [inline src/main/scala/akka/contrib/stream/pattern/reconnect/Reconnecter.scala]: `import context.dispatcher`?

File: src/main/scala/akka/contrib/stream/pattern/reconnect/ReconnectingStreamHttp.scala [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+  // format: OFF
+  def outgoingConnection(host: String,
+                         port: Int, reconnectInterval: FiniteDuration,
+                         settings: Option[ClientConnectionSettings],
+                         maxRetries: Long = Long.MaxValue)
+                        (onConnection: HttpConnected => Unit)
+                        (implicit sys: ActorSystem, mat: FlowMaterializer): Future[HttpConnected] = {
+    // format: ON
+    require(maxRetries >= 0, "maxRetries must be >= 0")
+    implicit val dispatcher = sys.dispatcher
```
Comment (hseeberger) [inline src/main/scala/akka/contrib/stream/pattern/reconnect/ReconnectingStreamHttp.scala]: `import sys.dispatcher`?

File: src/main/scala/akka/contrib/stream/pattern/reconnect/ReconnectingStreamHttp.scala [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+    val data = HttpReconnectionData(new InetSocketAddress(host, port), reconnectInterval, maxRetries, settings, onConnection)
+    implicit val timeout = Timeout(data.interval)
+
+    (sys.actorOf(Reconnecter.props(data, mat)) ? InitialConnect).mapTo[HttpConnected]
+  }
+
+  private class HttpReconnecter(initialData: HttpReconnectionData, implicit val mat: FlowMaterializer)
+      extends Reconnecter[HttpConnected, HttpReconnectionData](initialData) {
+    implicit val sched = context.dispatcher
```
Comment (hseeberger) [inline src/main/scala/akka/contrib/stream/pattern/reconnect/ReconnectingStreamHttp.scala]: Just import?

File: src/main/scala/akka/contrib/stream/pattern/reconnect/ReconnectingStreamHttp.scala [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+    implicit val sys = context.system
```
Comment (hseeberger) [inline src/main/scala/akka/contrib/stream/pattern/reconnect/ReconnectingStreamHttp.scala]: Just import?

File: src/main/scala/akka/contrib/stream/pattern/reconnect/ReconnectingStreamTcp.scala [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+   * The retutrned future will be completed for the initial connection,
+   * for subsequent reconnections (in case of stream failure etc) the given `onConnection` callback will be called.
+   */
+  // format: OFF
+  def outgoingConnection[M](address: InetSocketAddress, handleWith: Flow[ByteString, ByteString, M], reconnectInterval: FiniteDuration, maxRetries: Long = Long.MaxValue)
+                           (onConnection: (TcpConnected[M] => Unit))
+                           (implicit sys: ActorSystem, mat: FlowMaterializer): Future[TcpConnected[M]] = {
+    // format: ON
+    require(maxRetries >= 0, "maxRetries must be >= 0")
+    implicit val dispatcher = sys.dispatcher
```
Comment (hseeberger) [inline src/main/scala/akka/contrib/stream/pattern/reconnect/ReconnectingStreamTcp.scala]: Same here ...

File: src/main/scala/akka/contrib/stream/pattern/reconnect/ReconnectingStreamTcp.scala [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+
+    val data = TcpReconnectionData[M](address, handleWith, reconnectInterval, maxRetries, onConnection)
+    implicit val timeout = Timeout(data.interval)
```
Comment (hseeberger) [inline src/main/scala/akka/contrib/stream/pattern/reconnect/ReconnectingStreamTcp.scala]: Same here ;-)

File: src/main/scala/akka/contrib/stream/pattern/reconnect/ReconnectingStreamTcp.scala [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+
+    (sys.actorOf(Props(new TcpReconnectingActor(data, mat))) ? InitialConnect).mapTo[TcpConnected[M]]
+  }
+
+  private class TcpReconnectingActor[M](initialData: TcpReconnectionData[M], implicit val mat: FlowMaterializer)
+      extends Reconnecter[TcpConnected[M], ReconnectionData[TcpConnected[M]]](initialData) {
+
+    implicit val sched = context.dispatcher
```
Comment (hseeberger) [inline src/main/scala/akka/contrib/stream/pattern/reconnect/ReconnectingStreamTcp.scala]: And here (2 lines)

File: src/main/scala/akka/contrib/stream/pattern/reconnect/ReconnectingStreamTcp.scala [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+    }
+
+    /** Schedules an reconnect event when the upstream finishes (propagating the finishing) */
+    final class ReconnectStage[Mat](data: ReconnectionData[Mat])(implicit sys: ActorSystem, mat: FlowMaterializer) extends PushStage[ByteString, ByteString] {
+
+      override def onPush(elem: ByteString, ctx: Context[ByteString]): Directive = ctx.push(elem)
+
+      override def onUpstreamFailure(cause: Throwable, ctx: Context[ByteString]): TerminationDirective = {
+        if (data.retriesRemaining > 0) {
+          log.warning("Connection to {} was closed abruptly, reconnecting! (retries remaining: {})", data.address, data.retriesRemaining - 1)
```
Comment (hseeberger) [inline src/main/scala/akka/contrib/stream/pattern/reconnect/ReconnectingStreamTcp.scala]: "info" level here enough?

File: src/main/scala/akka/contrib/stream/pattern/reconnect/Reconnecter.scala (hunk 1 above)
Comment (ktoso) [inline src/main/scala/akka/contrib/stream/pattern/reconnect/Reconnecter.scala]: Both seem to be correct but I don't have a strong preference, Reconnector is is

**Comment by hseeberger** [comment]
LGTM apart from some nitpicks.

File: src/main/scala/akka/contrib/stream/pattern/reconnect/Reconnecter.scala (hunk 2 above)
Comment (ktoso) [inline src/main/scala/akka/contrib/stream/pattern/reconnect/Reconnecter.scala]: sure, okey

File: src/main/scala/akka/contrib/stream/pattern/reconnect/ReconnectingStreamTcp.scala (hunk 9 above)
Comment (ktoso) [inline src/main/scala/akka/contrib/stream/pattern/reconnect/ReconnectingStreamTcp.scala]: okey, the below one stays warning though - makes sense to me

**Pull request merged**